core/src/Streamly/Internal/Data/Array/Generic.hs

```diff
@@ -256,20 +256,23 @@ writeLastN n
         rb <- liftIO $ RB.createRing n
         return $ FL.Partial $ Tuple' rb (0 :: Int)
 
-    step (Tuple' rb rh) x = do
-        liftIO $ RB.unsafeInsertRing rb rh x
-        return $ FL.Partial $ Tuple' rb (rh + 1)
-
-    done (Tuple' rb rh) = do
-        arr' <- liftIO $ MArray.new (min rh n)
-        ref <- liftIO $ readIORef $ RB.ringHead rb
-        if rh < n
-        then
-            MArray.putSliceUnsafe (RB.arr rb) 0 arr' 0 ref
+    step (Tuple' rb cnt) x = do
+        liftIO $ RB.unsafeInsertRing rb x
+        return $ FL.Partial $ Tuple' rb (cnt + 1)
+
+    done (Tuple' rb cnt) = do
+        let len = min cnt n
+        arr <- liftIO $ MArray.new len
+        arr1 <- MArray.uninit arr len
+        -- XXX Should use an API for retrieving the ring head
+        rhead <- liftIO $ readIORef $ RB.ringHead rb
+        if cnt <= n
```

Comment:
Only this condiation has changed from the previous code, the rest is just cleanup.